lib/libc/rpc/rpc.3
.\"	$OpenBSD: rpc.3,v 1.50 2025/06/13 18:34:00 schwarze Exp $
.\"
.\" Copyright (c) 1998 Theo de Raadt
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" Copyright (c) 2010, Oracle America, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"     * Redistributions in binary form must reproduce the above
.\"       copyright notice, this list of conditions and the following
.\"       disclaimer in the documentation and/or other materials
.\"       provided with the distribution.
.\"     * Neither the name of the "Oracle America, Inc." nor the names of its
.\"       contributors may be used to endorse or promote products derived
.\"       from this software without specific prior written permission.
.\"
.\"   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\"   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\"   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
.\"   FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
.\"   COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
.\"   INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\"   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
.\"   GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\"   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\"   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\"   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\"   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: June 13 2025 $
.Dt CALLRPC 3
.Os
.Sh NAME
.Nm callrpc ,
.Nm clnt_broadcast ,
.Nm clnt_call ,
.Nm clnt_control ,
.Nm clnt_create ,
.Nm clnt_destroy ,
.Nm clnt_freeres ,
.Nm clnt_pcreateerror ,
.Nm clnt_perrno ,
.Nm clnt_perror ,
.Nm clnt_spcreateerror ,
.Nm clnt_sperrno ,
.Nm clnt_sperror ,
.Nm clntraw_create ,
.Nm clnttcp_create ,
.Nm clntudp_bufcreate ,
.Nm clntudp_create ,
.Nm clnt_geterr ,
.Nm get_myaddress ,
.Nm pmap_getmaps ,
.Nm pmap_getport ,
.Nm pmap_rmtcall ,
.Nm pmap_set ,
.Nm pmap_unset ,
.Nm registerrpc ,
.Nm rpc_createerr ,
.Nm svc_destroy ,
.Nm svc_fds ,
.Nm svc_fdset ,
.Nm svc_freeargs ,
.Nm svc_getargs ,
.Nm svc_getcaller ,
.Nm svc_getreq ,
.Nm svc_getreq_common ,
.Nm svc_getreq_poll ,
.Nm svc_getreqset ,
.Nm svc_getreqset2 ,
.Nm svc_register ,
.Nm svc_max_pollfd ,
.Nm svc_pollfd ,
.Nm svc_run ,
.Nm svc_sendreply ,
.Nm svc_unregister ,
.Nm svcerr_auth ,
.Nm svcerr_decode ,
.Nm svcerr_noproc ,
.Nm svcerr_noprog ,
.Nm svcerr_progvers ,
.Nm svcerr_systemerr ,
.Nm svcerr_weakauth ,
.Nm svcfd_create ,
.Nm svcraw_create ,
.Nm svctcp_create ,
.Nm svcudp_create ,
.Nm svcudp_bufcreate ,
.Nm xdr_accepted_reply ,
.Nm xdr_authunix_parms ,
.Nm xdr_callhdr ,
.Nm xdr_callmsg ,
.Nm xdr_opaque_auth ,
.Nm xdr_pmap ,
.Nm xdr_pmaplist ,
.Nm xdr_rejected_reply ,
.Nm xdr_replymsg ,
.Nm xprt_register ,
.Nm xprt_unregister
.Nd library routines for remote procedure calls
.Sh SYNOPSIS
.In rpc/rpc.h
.Ft int
.Fn callrpc "char *host" "u_long prognum" "u_long versnum" "u_long procnum" "xdrproc_t inproc" "char *in" "xdrproc_t outproc" "char *out"
.Ft enum clnt_stat
.Fn clnt_broadcast "u_long prognum" "u_long versnum" "u_long procnum" "xdrproc_t inproc" "char *in" "xdrproc_t outproc" "char *out" "resultproc_t eachresult"
.Ft enum clnt_stat
.Fn clnt_call "CLIENT *clnt" "u_long procnum" "xdrproc_t inproc" "char *in" "xdrproc_t outproc" "char *out" "struct timeval tout"
.Ft int
.Fn clnt_destroy "CLIENT *clnt"
.Ft CLIENT *
.Fn clnt_create "char *host" "u_long prog" "u_long vers" "char *proto"
.Ft bool_t
.Fn clnt_control "CLIENT *cl" "int req" "char *info"
.Ft int
.Fn clnt_freeres "CLIENT *clnt" "xdrproc_t outproc" "char *out"
.Ft void
.Fn clnt_geterr "CLIENT *clnt" "struct rpc_err *errp"
.Ft void
.Fn clnt_pcreateerror "char *s"
.Ft void
.Fn clnt_perrno "enum clnt_stat stat"
.Ft int
.Fn clnt_perror "CLIENT *clnt" "char *s"
.Ft char *
.Fn clnt_spcreateerror "char *s"
.Ft char *
.Fn clnt_sperrno "enum clnt_stat stat"
.Ft char *
.Fn clnt_sperror "CLIENT *rpch" "char *s"
.Ft CLIENT *
.Fn clntraw_create "u_long prognum" "u_long versnum"
.Ft CLIENT *
.Fn clnttcp_create "struct sockaddr_in *addr" "u_long prognum" "u_long versnum" "int *sockp" "u_int sendsz" "u_int recvsz"
.Ft CLIENT *
.Fn clntudp_create "struct sockaddr_in *addr" "u_long prognum" "u_long versnum" "struct timeval wait" "int *sockp"
.Ft CLIENT *
.Fn clntudp_bufcreate "struct sockaddr_in *addr" "u_long prognum" "u_long versnum" "struct timeval wait" "int *sockp" "unsigned int sendsize" "unsigned int recosize"
.Ft int
.Fn get_myaddress "struct sockaddr_in *addr"
.Ft struct pmaplist *
.Fn pmap_getmaps "struct sockaddr_in *addr"
.Ft u_short
.Fn pmap_getport "struct sockaddr_in *addr" "u_long prognum" "u_long versnum" "u_long protocol"
.Ft enum clnt_stat
.Fn pmap_rmtcall "struct sockaddr_in *" "u_long prog" "u_long vers" "u_long proc" "xdrproc_t inp" "char *in" "xdrproc_t outp" "char *out" "struct timeval tv" "u_long *portp"
.Ft int
.Fn pmap_set "u_long prognum" "u_long versnum" "u_int protocol" "int port"
.Ft int
.Fn pmap_unset "u_long prognum" "u_long versnum"
.Ft int
.Fn registerrpc "u_long prognum" "u_long versnum" "u_long procnum" "char *(*procname)() " "xdrproc_t inproc" "xdrproc_t outproc"
.Vt struct rpc_createerr rpc_createerr ;
.Ft int
.Fn svc_destroy "SVCXPRT *xprt"
.Vt struct pollfd *svc_pollfd ;
.Vt int svc_max_pollfd ;
.Vt fd_set svc_fdset ;
.Vt fd_set *__svc_fdset ;
.Vt int __svc_fdsetsize ;
.Vt int svc_fds ;
.Ft int
.Fn svc_freeargs "SVCXPRT *xprt" "xdrproc_t inproc" "char *in"
.Ft int
.Fn svc_getargs "SVCXPRT *xprt" "xdrproc_t inproc" "char *in"
.Ft struct sockaddr_in *
.Fn svc_getcaller "SVCXPRT *xprt"
.Ft int
.Fn svc_getreq_common "int fd"
.Ft int
.Fn svc_getreq_poll "struct pollfd *pfds" "const int pollretval"
.Ft int
.Fn svc_getreqset "fd_set *rdfds"
.Ft int
.Fn svc_getreqset2 "fd_set *rdfds" "int width"
.Ft int
.Fn svc_getreq "int rdfds"
.Ft int
.Fn svc_register "SVCXPRT *xprt" "u_long prognum" "u_long versnum" "void (*dispatch)()" "u_long protocol"
.Ft int
.Fn svc_run "void"
.Ft int
.Fn svc_sendreply "SVCXPRT *xprt" "xdrproc_t outproc" "char *out"
.Ft void
.Fn svc_unregister "u_long prognum" "u_long versnum"
.Ft void
.Fn svcerr_auth "SVCXPRT *xprt" "enum auth_stat why"
.Ft void
.Fn svcerr_decode "SVCXPRT *xprt"
.Ft void
.Fn svcerr_noproc "SVCXPRT *xprt"
.Ft void
.Fn svcerr_noprog "SVCXPRT *xprt"
.Ft void
.Fn svcerr_progvers "SVCXPRT *xprt"
.Ft void
.Fn svcerr_systemerr "SVCXPRT *xprt"
.Ft void
.Fn svcerr_weakauth "SVCXPRT *xprt"
.Ft SVCXPRT *
.Fn svcraw_create "void"
.Ft SVCXPRT *
.Fn svctcp_create "int sock" "u_int send_buf_size" "u_int recv_buf_size"
.Ft SVCXPRT *
.Fn svcfd_create "int fd" "u_int sendsize" "u_int recvsize"
.Ft SVCXPRT *
.Fn svcudp_create "int sock"
.Ft SVCXPRT *
.Fn svcudp_bufcreate "int sock" "u_int sendsz" "u_int recvsz"
.Ft bool_t
.Fn xdr_accepted_reply "XDR *xdrs" "struct accepted_reply *ar"
.Ft bool_t
.Fn xdr_authunix_parms "XDR *xdrs" "struct authunix_parms *aupp"
.Ft void
.Fn xdr_callhdr "XDR *xdrs" "struct rpc_msg *chdr"
.Ft int
.Fn xdr_callmsg "XDR *xdrs" "struct rpc_msg *cmsg"
.Ft int
.Fn xdr_opaque_auth "XDR *xdrs" "struct opaque_auth *ap"
.Ft int
.Fn xdr_pmap "XDR *xdrs" "struct pmap *regs"
.Ft int
.Fn xdr_pmaplist "XDR *xdrs" "struct pmaplist **rp"
.Ft int
.Fn xdr_rejected_reply "XDR *xdrs" "struct rejected_reply *rr"
.Ft int
.Fn xdr_replymsg "XDR *xdrs" "struct rpc_msg *rmsg"
.Ft void
.Fn xprt_register "SVCXPRT *xprt"
.Ft void
.Fn xprt_unregister "SVCXPRT *xprt"
.Sh DESCRIPTION
These routines allow C programs to make procedure
calls on other machines across the network.
First, the client calls a procedure to send a
data packet to the server.
Upon receipt of the packet, the server calls a dispatch routine
to perform the requested service, and then sends back a
reply.
Finally, the procedure call returns to the client.
.Pp
.\"Routines that are used for Secure RPC (DES authentication) are described in
.\".Xr rpc_secure 3 .
.\"Secure RPC can be used only if DES encryption is available.
.Fn callrpc
calls the remote procedure associated with
.Fa prognum ,
.Fa versnum ,
and
.Fa procnum
on the machine,
.Fa host .
The parameter
.Fa in
is the address of the procedure's argument(s), and
.Fa out
is the address of where to place the result(s);
.Fa inproc
is used to encode the procedure's parameters, and
.Fa outproc
is used to decode the procedure's results.
This routine returns zero if it succeeds, or the value of
.Fa enum clnt_stat
cast to an integer if it fails.
The routine
.Fn clnt_perrno
is handy for translating failure statuses into messages.
.Pp
.Sy Warning:
calling remote procedures with this routine uses UDP/IP
as a transport; see
.Fn clntudp_create
for restrictions.
You do not have control of timeouts or authentication using
this routine.
.Pp
.Fn clnt_broadcast
is like
.Fn callrpc ,
except the call message is broadcast to all locally
connected broadcast nets.
Each time it receives a response, this routine calls
.Fa eachresult ,
whose form is:
.Bd -literal -offset indent
.Ft int
.Fn eachresult "char *out" "struct sockaddr_in *addr"
.Ed
.Pp
where
.Fa out
is the same as
.Fa out
passed to
.Fn clnt_broadcast ,
except that the remote procedure's output is decoded there;
.Fa addr
points to the address of the machine that sent the results.
If
.Fa eachresult
returns zero,
.Fn clnt_broadcast
waits for more replies; otherwise it returns with appropriate
status.
.Pp
.Sy Warning:
broadcast sockets are limited in size to the
maximum transfer unit of the data link.
For Ethernet, this value is 1500 bytes.
.Pp
.Fn clnt_call
is a macro that calls the remote procedure
.Fa procnum
associated with the client handle,
.Fa clnt ,
which is obtained with an RPC client creation routine such as
.Fn clnt_create .
The parameter
.Fa in
is the address of the procedure's argument(s), and
.Fa out
is the address of where to place the result(s);
.Fa inproc
is used to encode the procedure's parameters, and
.Fa outproc
is used to decode the procedure's results;
.Fa tout
is the time allowed for results to come back.
.Pp
.Fn clnt_destroy
is a macro that destroys the client's RPC handle.
Destruction usually involves deallocation of private data structures, including
.Fa clnt
itself.
Use of
.Fa clnt
is undefined after calling
.Fn clnt_destroy .
If the RPC library opened the associated socket, it will close it also.
Otherwise, the socket remains open.
.Pp
.Fn clnt_create
is a generic client creation routine.
.Fa host
identifies the name of the remote host where the server
is located.
.Fa proto
indicates which kind of transport protocol to use.
The currently supported values for this field are
.Qq udp
and
.Qq tcp .
Default timeouts are set, but can be modified using
.Fn clnt_control .
This routine returns
.Dv NULL
if it fails.
.Pp
.Sy Warning:
Using UDP has its shortcomings.
Since UDP-based RPC
messages can only hold up to 8 Kbytes of encoded data,
this transport cannot be used for procedures that take
large arguments or return huge results.
.Pp
.Fn clnt_control
is a macro used to change or retrieve various information
about a client object.
.Fa req
indicates the type of operation, and
.Fa info
is a pointer to the information.
For both UDP and TCP,
the supported values of
.Fa req
and their argument types and what they do are:
.Bd -literal -offset indent
CLSET_TIMEOUT	struct timeval	set total timeout
CLGET_TIMEOUT	struct timeval	get total timeout
.Ed
.Pp
.Sy Note:
if you set the timeout using
.Fn clnt_control ,
the timeout parameter passed to
.Fn clnt_call
will be ignored in all future calls.
.Bd -literal -offset indent
CLGET_SERVER_ADDR	struct sockaddr_in 	get server's address
.Ed
.Pp
The following operations are valid for UDP only:
.Bd -literal -offset indent
CLSET_RETRY_TIMEOUT   struct timeval	set the retry timeout
CLGET_RETRY_TIMEOUT   struct timeval	get the retry timeout
.Ed
.Pp
The retry timeout is the time that UDP RPC
waits for the server to reply before
retransmitting the request.
.Pp
.Fn clnt_freeres
is a macro that frees any data allocated by the RPC/XDR
system when it decoded the results of an RPC call.
The parameter
.Fa out
is the address of the results, and
.Fa outproc
is the XDR routine describing the results.
This routine returns one if the results were successfully
freed,
and zero otherwise.
.Pp
.Fn clnt_geterr
is a macro that copies the error structure out of the client
handle
to the structure at address
.Fa errp .
.Pp
.Fn clnt_pcreateerror
prints a message to standard error indicating
why a client RPC handle could not be created.
The message is prepended with string
.Fa s
and a colon.
Used when a
.Fn clnt_create ,
.Fn clntraw_create ,
.Fn clnttcp_create ,
or
.Fn clntudp_create
call fails.
.Pp
.Fn clnt_perrno
prints a message to standard error corresponding
to the condition indicated by
.Fa stat .
Used after
.Fn callrpc .
.Pp
.Fn clnt_perror
prints a message to standard error indicating why an RPC call failed;
.Fa clnt
is the handle used to do the call.
The message is prepended with string
.Fa s
and a colon.
Used after
.Fn clnt_call .
.Pp
.Fn clnt_spcreateerror
is like
.Fn clnt_pcreateerror ,
except that it returns a string
instead of printing to the standard error.
.Pp
.Sy Bugs:
returns pointer to static data that is overwritten
on each call.
.Pp
.Fn clnt_sperrno
takes the same arguments as
.Fn clnt_perrno ,
but instead of sending a message to the standard error
indicating why an RPC
call failed, returns a pointer to a string which contains
the message.
Unlike
.Fn clnt_perror ,
it does not append a newline character
to the end of the message.
.Pp
.Fn clnt_sperrno
is used instead of
.Fn clnt_perrno
if the program does not have a standard error (as a program
running as a server quite likely does not), or if the
programmer
does not want the message to be output with
.Fn printf ,
or if a message format different than that supported by
.Fn clnt_perrno
is to be used.
.Pp
.Sy Note:
unlike
.Fn clnt_sperror
and
.Fn clnt_spcreaterror ,
.Fn clnt_sperrno
returns a pointer to static data, but the
result will not get overwritten on each call.
.Pp
.Fn clnt_sperror
is like
.Fn clnt_perror ,
except that (like
.Fn clnt_sperrno )
it returns a string instead of printing to standard error.
.Pp
.Sy Bugs:
returns pointer to static data that is overwritten
on each call.
.Pp
.Fn clntraw_create
is a routine which creates a toy RPC client for the remote program
.Fa prognum ,
version
.Fa versnum .
The transport used to pass messages to the service is
actually a buffer within the process's address space, so the
corresponding RPC server should live in the same address space; see
.Fn svcraw_create .
This allows simulation of RPC and acquisition of RPC
overheads, such as round trip times, without any
kernel interference.
This routine returns
.Dv NULL
if it fails.
.Pp
.Fn clnttcp_create
is a routine which creates an RPC client for the remote program
.Fa prognum ,
version
.Fa versnum ;
the client uses TCP/IP as a transport.
The remote program is located at Internet address
.Fa *addr .
If
.Fa addr-\*(Gtsin_port
is zero, then it is set to the actual port that the remote
program is listening on (the remote
.Xr portmap 8
service is consulted for this information).
The parameter
.Fa sockp
is a socket; if it is
.Fa RPC_ANYSOCK ,
then this routine opens a new one and sets
.Fa sockp .
Since TCP-based RPC uses buffered I/O,
the user may specify the size of the send and receive buffers
with the parameters
.Fa sendsz
and
.Fa recvsz ;
values of zero choose suitable defaults.
This routine returns
.Dv NULL
if it fails.
.Pp
.Fn clntudp_create
is a routine which creates an RPC client for the remote program
.Fa prognum ,
on
.Fa versnum ;
the client uses use UDP/IP as a transport.
The remote program is located at Internet address
.Fa addr .
If
.Fa addr-\*(Gtsin_port
is zero, then it is set to actual port that the remote
program is listening on (the remote
.Xr portmap 8
service is consulted for this information).
The parameter
.Fa sockp
is a socket; if it is
.Fa RPC_ANYSOCK ,

debian/changelog
linux (6.12.10-76061203.202412060638) oracular; urgency=low

  [ Mainline Build ]
  Mainline build at commit: v6.12.3

 -- Michael Aaron Murphy <michael@mmurphy.dev>  Mon, 20 Jan 2025 09:42:56 +0100


sys/lib/libsa/rpc.c
/*	$OpenBSD: rpc.c,v 1.15 2015/08/15 19:42:56 miod Exp $	*/
/*	$NetBSD: rpc.c,v 1.16 1996/10/13 02:29:06 christos Exp $	*/

/*
 * Copyright (c) 1992 Regents of the University of California.
 * All rights reserved.
 *
 * This software was developed by the Computer Systems Engineering group
 * at Lawrence Berkeley Laboratory under DARPA contract BG 91-66 and
 * contributed to Berkeley.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Lawrence Berkeley Laboratory and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * @(#) Header: rpc.c,v 1.12 93/09/28 08:31:56 leres Exp  (LBL)
 */

/*
 * RPC functions used by NFS and bootparams.
 * Note that bootparams requires the ability to find out the
 * address of the server from which its response has come.
 * This is supported by keeping the IP/UDP headers in the
 * buffer space provided by the caller.  (See rpc_fromaddr)
 */

#include <sys/param.h>
#include <sys/socket.h>

#include <netinet/in.h>

#include <nfs/rpcv2.h>

#include "stand.h"
#include "net.h"
#include "netif.h"
#include "rpc.h"

struct auth_info {
	int32_t		authtype;	/* auth type */
	u_int32_t	authlen;	/* auth length */
};

struct auth_unix {
	int32_t   ua_time;
	int32_t   ua_hostname;	/* null */
	int32_t   ua_uid;
	int32_t   ua_gid;
	int32_t   ua_gidlist;	/* null */
};

struct rpc_call {
	u_int32_t	rp_xid;		/* request transaction id */
	int32_t		rp_direction;	/* call direction (0) */
	u_int32_t	rp_rpcvers;	/* rpc version (2) */
	u_int32_t	rp_prog;	/* program */
	u_int32_t	rp_vers;	/* version */
	u_int32_t	rp_proc;	/* procedure */
};

struct rpc_reply {
	u_int32_t	rp_xid;		/* request transaction id */
	int32_t		rp_direction;	/* call direction (1) */
	int32_t		rp_astatus;	/* accept status (0: accepted) */
	union {
		u_int32_t	rpu_errno;
		struct {
			struct auth_info rok_auth;
			u_int32_t	rok_status;
		} rpu_rok;
	} rp_u;
};

/* Local forwards */
static	ssize_t recvrpc(struct iodesc *, void *, size_t, time_t);
static	int rpc_getport(struct iodesc *, u_int32_t, u_int32_t);

int rpc_xid;
int rpc_port = 0x400;	/* predecrement */

/*
 * Make a rpc call; return length of answer
 * Note: Caller must leave room for headers.
 */
ssize_t
rpc_call(struct iodesc *d, u_int32_t prog, u_int32_t vers, u_int32_t proc, void *sdata,
    size_t slen, void *rdata, size_t rlen)
{
	ssize_t cc;
	struct auth_info *auth;
	struct rpc_call *call;
	struct rpc_reply *reply;
	char *send_head, *send_tail;
	char *recv_head, *recv_tail;
	u_int32_t x;
	int port;	/* host order */

#ifdef RPC_DEBUG
	if (debug)
		printf("rpc_call: prog=0x%x vers=%d proc=%d\n",
		    prog, vers, proc);
#endif

	port = rpc_getport(d, prog, vers);
	if (port == -1)
		return (-1);

	d->destport = htons(port);

	/*
	 * Prepend authorization stuff and headers.
	 * Note, must prepend things in reverse order.
	 */
	send_head = sdata;
	send_tail = (char *)sdata + slen;

	/* Auth verifier is always auth_null */
	send_head -= sizeof(*auth);
	auth = (struct auth_info *)send_head;
	auth->authtype = htonl(RPCAUTH_NULL);
	auth->authlen = 0;

#if 1
	/* Auth credentials: always auth unix (as root) */
	send_head -= sizeof(struct auth_unix);
	bzero(send_head, sizeof(struct auth_unix));
	send_head -= sizeof(*auth);
	auth = (struct auth_info *)send_head;
	auth->authtype = htonl(RPCAUTH_UNIX);
	auth->authlen = htonl(sizeof(struct auth_unix));
#else
	/* Auth credentials: always auth_null (XXX OK?) */
	send_head -= sizeof(*auth);
	auth = send_head;
	auth->authtype = htonl(RPCAUTH_NULL);
	auth->authlen = 0;
#endif

	/* RPC call structure. */
	send_head -= sizeof(*call);
	call = (struct rpc_call *)send_head;
	rpc_xid++;
	call->rp_xid       = htonl(rpc_xid);
	call->rp_direction = htonl(RPC_CALL);
	call->rp_rpcvers   = htonl(RPC_VER2);
	call->rp_prog = htonl(prog);
	call->rp_vers = htonl(vers);
	call->rp_proc = htonl(proc);

	/* Make room for the rpc_reply header. */
	recv_head = rdata;
	recv_tail = (char *)rdata + rlen;
	recv_head -= sizeof(*reply);

	cc = sendrecv(d,
	    sendudp, send_head, send_tail - send_head,
	    recvrpc, recv_head, recv_tail - recv_head);

#ifdef RPC_DEBUG
	if (debug)
		printf("callrpc: cc=%d rlen=%d\n", cc, rlen);
#endif
	if (cc <= -1)
		return (-1);

	if ((size_t)cc <= sizeof(*reply)) {
		errno = EBADRPC;
		return (-1);
	}

	recv_tail = recv_head + cc;

	/*
	 * Check the RPC reply status.
	 * The xid, dir, astatus were already checked.
	 */
	reply = (struct rpc_reply *)recv_head;
	auth = &reply->rp_u.rpu_rok.rok_auth;
	x = ntohl(auth->authlen);
	if (x != 0) {
#ifdef RPC_DEBUG
		if (debug)
			printf("callrpc: reply auth != NULL\n");
#endif
		errno = EBADRPC;
		return(-1);
	}
	x = ntohl(reply->rp_u.rpu_rok.rok_status);
	if (x != 0) {
		printf("callrpc: error = %d\n", x);
		errno = EBADRPC;
		return(-1);
	}
	recv_head += sizeof(*reply);

	return (ssize_t)(recv_tail - recv_head);
}

/*
 * Returns true if packet is the one we're waiting for.
 * This just checks the XID, direction, acceptance.
 * Remaining checks are done by callrpc
 */
static ssize_t
recvrpc(struct iodesc *d, void *pkt, size_t len, time_t tleft)
{
	struct rpc_reply *reply;
	ssize_t	n;
	int	x;

	errno = 0;
#ifdef RPC_DEBUG
	if (debug)
		printf("recvrpc: called len=%d\n", len);
#endif

	n = readudp(d, pkt, len, tleft);
	if (n <= (4 * 4))
		return -1;

	reply = (struct rpc_reply *)pkt;

	x = ntohl(reply->rp_xid);
	if (x != rpc_xid) {
#ifdef RPC_DEBUG
		if (debug)
			printf("recvrpc: rp_xid %d != xid %d\n", x, rpc_xid);
#endif
		return -1;
	}

	x = ntohl(reply->rp_direction);
	if (x != RPC_REPLY) {
#ifdef RPC_DEBUG
		if (debug)
			printf("recvrpc: rp_direction %d != REPLY\n", x);
#endif
		return -1;
	}

	x = ntohl(reply->rp_astatus);
	if (x != RPC_MSGACCEPTED) {
		errno = ntohl(reply->rp_u.rpu_errno);
		printf("recvrpc: reject, astat=%d, errno=%d\n", x, errno);
		return -1;
	}

	/* Return data count (thus indicating success) */
	return (n);
}

/*
 * Given a pointer to a reply just received,
 * dig out the IP address/port from the headers.
 */
void
rpc_fromaddr(void *pkt, struct in_addr *addr, u_short *port)
{
	struct hackhdr {
		/* Tail of IP header: just IP addresses */
		u_int32_t ip_src;
		u_int32_t ip_dst;
		/* UDP header: */
		u_int16_t uh_sport;		/* source port */
		u_int16_t uh_dport;		/* destination port */
		int16_t	  uh_ulen;		/* udp length */
		u_int16_t uh_sum;		/* udp checksum */
		/* RPC reply header: */
		struct rpc_reply rpc;
	} *hhdr;

	hhdr = ((struct hackhdr *)pkt) - 1;
	addr->s_addr = hhdr->ip_src;
	*port = hhdr->uh_sport;
}

/*
 * RPC Portmapper cache
 */
#define PMAP_NUM 8			/* need at most 5 pmap entries */

int rpc_pmap_num;
struct pmap_list {
	struct in_addr	addr;	/* server, net order */
	u_int	prog;		/* host order */
	u_int	vers;		/* host order */
	int	port;		/* host order */
} rpc_pmap_list[PMAP_NUM];

/* return port number in host order, or -1 */
int
rpc_pmap_getcache(struct in_addr addr, u_int prog, u_int vers)
{
	struct pmap_list *pl;

	for (pl = rpc_pmap_list; pl < &rpc_pmap_list[rpc_pmap_num]; pl++) {
		if (pl->addr.s_addr == addr.s_addr &&
		    pl->prog == prog && pl->vers == vers)
			return (pl->port);
	}
	return (-1);
}

void
rpc_pmap_putcache(struct in_addr addr, u_int prog, u_int vers, int port)
{
	struct pmap_list *pl;

	/* Don't overflow cache... */
	if (rpc_pmap_num >= PMAP_NUM) {
		/* ... just re-use the last entry. */
		rpc_pmap_num = PMAP_NUM - 1;
#ifdef	RPC_DEBUG
		printf("rpc_pmap_putcache: cache overflow\n");
#endif
	}

	pl = &rpc_pmap_list[rpc_pmap_num];
	rpc_pmap_num++;

	/* Cache answer */
	pl->addr = addr;
	pl->prog = prog;
	pl->vers = vers;
	pl->port = port;
}


/*
 * Request a port number from the port mapper.
 * Returns the port in host order.
 */
int
rpc_getport(struct iodesc *d, u_int32_t prog, u_int32_t vers)
{
	struct args {
		u_int32_t	prog;		/* call program */
		u_int32_t	vers;		/* call version */
		u_int32_t	proto;		/* call protocol */
		u_int32_t	port;		/* call port (unused) */
	} *args;
	struct res {
		u_int32_t port;
	} *res;
	struct {
		u_int32_t	h[RPC_HEADER_WORDS];
		struct args d;
	} sdata;
	struct {
		u_int32_t	h[RPC_HEADER_WORDS];
		struct res d;
		u_int32_t  pad;
	} rdata;
	ssize_t cc;
	int port;

#ifdef RPC_DEBUG
	if (debug)
		printf("getport: prog=0x%x vers=%d\n", prog, vers);
#endif

	/* This one is fixed forever. */
	if (prog == PMAPPROG)
		return (PMAPPORT);

	/* Try for cached answer first */
	port = rpc_pmap_getcache(d->destip, prog, vers);
	if (port != -1)
		return (port);

	args = &sdata.d;
	args->prog = htonl(prog);
	args->vers = htonl(vers);
	args->proto = htonl(IPPROTO_UDP);
	args->port = 0;
	res = &rdata.d;

	cc = rpc_call(d, PMAPPROG, PMAPVERS, PMAPPROC_GETPORT,
		args, sizeof(*args), res, sizeof(*res));
	if (cc < 0 || (size_t)cc < sizeof(*res)) {
		printf("getport: %s", strerror(errno));
		errno = EBADRPC;
		return (-1);
	}
	port = (int)ntohl(res->port);

	rpc_pmap_putcache(d->destip, prog, vers, port);

	return (port);
}
